UDP query of game script

TheDude-gh opened the pull request:
Reopening the idea of UDP query of game script and probably showing it also in game GUI.

This patch focuses to extend fetching game info over UDP from multiplayer servers. Currently you can get basic game information and NewGRF details, but nothing about game scripts.

I think it would be useful to have possibility to found out, if server is running game scripts. When servers has a game scripts, it and could be shown in ingame lobby window and could be added also to web server listing .

Currently this patch solves it by adding new packet type, so there is extra UDP communication needed, but I think it would be better to rewrite it a little bit and incresase NETWORK_GAME_INFO_VERSION and add it to the end of the PACKET_SERVER_GAME_INFO packet in SerializeNetworkGameInfo() in game_info.cpp.

TheDude-gh committed:
Add: UDP query of game script

Reopened from old flyspray issue #6211.

Patch adds a possibility to query OpenTTD server for running game scripts and returns its name and version if any present.

Goal of this fork is to make also ingame interface.

TheDude-gh committed:
Merge branch 'master' of https://github.com/OpenTTD/OpenTTD

TrueBrain commented:
Honestly, I am not really sure what to do with this.

You mention that you want to reopen an idea, but you don't start a discussion; you just show some code. You really have to explain a bit more.

The Pull Request template we have helps you with that. It tells you to write a motivation and a description of your Pull Request. But as you didn't do that, you didn't "reopen" anything really. You just showed us some code.

Please help us help you. Don't make it our work to figure out what your intentions are, what you want to accomplish, etc. That really is your job: convince us why we should want this. And for that we have a template, to guide you through that process.

Could you please fill in the template and help us out what is the idea here?

Thank you!

TheDude-gh commented:
I have probably missed the template, my bad.

This patch focuses to extend fetching game info over UDP from multiplayer servers. Currently you can get basic game information and NewGRF details, but nothing about game scripts.

I think it would be useful to have possibility to found out, if server is running game scripts. When servers has a game scripts, it and could be shown in ingame lobby window and could be added also to web server listing .

Currently this patch solves it by adding new packet type, so there is extra UDP communication needed, but I think it would be better to rewrite it a little bit and incresase NETWORK_GAME_INFO_VERSION and add it to the end of the PACKET_SERVER_GAME_INFO packet in SerializeNetworkGameInfo() in game_info.cpp.

TrueBrain commented:
What I miss in your description, is a motivation. What do you want to accomplish with this?

I can see it adds a packet that gives some GameScript information. But why specific that information? And why that alone?

In other words: how do you intend to use it? What is the request that made you build this? Just that you find it useful is not really a convincing argument :) Please elaborate a bit more :) Tnx!

(edit: at least, I assume you didn't add a full packet just for the game script name; that could be done in the GameInfo too. Given the amount of data, I assume you have a higher purpose here. But this is my interpretation, so a bit more explaining how you want to use this would really help)

TheDude-gh commented:
I have two motivations:
1) When playing multiplayer, you could look if there is specific game script on some of the servers (that means a filter option for game scripts would have to be included in patch).
2) A search over all servers could be done to have statistics about game scripts used, a similar thing I have done for newgrfs some time ago.

TrueBrain commented:
Tnx!

In that case this PR needs a bit more work of course :)

Starters, I would add this to GameInfo. It is just a simple string, possibly with a version, so that would make two simple strings, or maybe merge it into a single one ("name version" or something). It means we don't need a new packet, which is always a bonus.

But mostly, this PR would also need to implement the usage of this packet. We tend not to like stuff that just adds something that someday someone might use .. those things tend to be forgotten, and break :P

Did you have any thought on how that would work in the GUI?

Personally, I am not against something like this; but this currently is just an idea that I guess needs to be worked out a bit further :)

PS: if we do this for GS, we might as well do it for AIs too, but that is another story :)

TheDude-gh commented:
My idea of GUI is to have new text about GS in the right pane with game info, when server is selected. 

And new filter option for game script somewhere in the area where current server name filter is.

![Schránka 02](https://user-images.githubusercontent.com/47122855/117711650-cff3a880-b1d3-11eb-8229-399d96d86ef9.png)

But I am not really sure, whether to add the filter right away. Maybe I would start with adding just the UDP info and the simple new row, and when it is included, we could see how much are GS actually used across multiplayer games. And just after that add new filter.
Depends on the opinions from development team.

As for AI information, I would'n mix it with this patch. Are AI's even used in multiplayer games? While I can easily imagine people would search for specific GS, I am not so convinced if anyone would be interested in server with AI's.

TheDude-gh committed:
UDP GS Query rewrite
New version of gameinfo packet with extended detail about gamescript instead of new packet type

TrueBrain commented on src/network/network_udp.cpp:
```diff
@@ -435,6 +436,34 @@ void ClientNetworkUDPSocketHandler::Receive_SERVER_NEWGRFS(Packet *p, NetworkAdd
 }
 
+
+void ClientNetworkUDPSocketHandler::HandleIncomingNetworkGameInfoGRFConfig(GRFConfig *config)
```
What does this function do? It is never called?

TheDude-gh replied:
It should not be there. I somehow screwed up fetching updates from the old fork.

TrueBrain commented on src/network/network_udp.cpp:
```diff
@@ -295,6 +295,7 @@ void ServerNetworkUDPSocketHandler::Receive_CLIENT_GET_NEWGRFS(Packet *p, Networ
 	this->SendPacket(&packet, client_addr);
 }
 
+
```
Extra newline :) Please keep a diff as small as possible :)

TheDude-gh replied:
OK

TrueBrain commented on src/network/network_gui.cpp:
```diff
@@ -660,6 +660,13 @@ class NetworkGameWindow : public Window {
+			if (sel->info.has_gamescript) {
+				SetDParamStr(0, sel->info.gs_name);
+				SetDParam(1, sel->info.gs_version);
+				DrawString(r.left + WD_FRAMERECT_LEFT, r.right - WD_FRAMERECT_RIGHT, y, STR_NETWORK_SERVER_LIST_GAMESCRIPT); // current date
```
That comment is a lie

TheDude-gh replied:
Will be fixed

TrueBrain commented on src/network/core/game_info.cpp:
```diff
@@ -198,6 +200,21 @@ void SerializeNetworkGameInfo(Packet *p, const NetworkServerGameInfo *info)
+		GameInfo* gsinfo = Game::GetInfo();
+		if (gsinfo == nullptr) {
+			p->Send_bool(false); //has gamescript?
+		}
```
I do not think that sending a bool is useful. Just an empty string is sufficient to spot there is no GS.

TheDude-gh replied:
Is there a difference? Both will send zero byte and bool(false) or uint8(0) seems more understandable to me.

TrueBrain replied:
Not sure we are speaking the same language here :)

My suggestion is to not send this bool at all. Just run: Send_string(name); Send_uint32(version);. And if those are "" and -1, you know there was not GS loaded. That is a lot simpler throughout the code, as it saves the constant checking if the bool was set yes/no :)

PS: did you know that for every comment you make on GitHub, I get a lovely email :D Not a real problem, but just so you know. So answer like "OK" and "will be fixed" .. I will see that next push, don't worry about it :)

TheDude-gh replied:
It save one check when reading received packet. A check for existing GS still has to be there, so you don't read properties of null and also check for GUI has to stay, it will just change to something like 

if(gs_version != -1) DrawString()...

I would consider having extra bool variable better for understanding code than arbitrary -1 value, but both are sufficient, so I guess, why not.

TrueBrain commented on src/network/core/game_info.cpp:
```diff
@@ -198,6 +200,21 @@ void SerializeNetworkGameInfo(Packet *p, const NetworkServerGameInfo *info)
+		else {
+			p->Send_bool(true); //has gamescript?
+			p->Send_uint32((uint32)gsinfo->GetVersion());
+			p->Send_string(gsinfo->GetShortName());
```
You send the shortname but never use it. I also don't think it is useful for anyone to know, really.

TheDude-gh replied:
At this moment no, it's not used. Short name is somewhat unique to each gamescript, so it might be used for looking the script in game content. For current patch it's useless, question would be if it can be used in future. Otherwise I agree it can be just omitted.

TrueBrain replied:
I doubt it will be useful, but as I said earlier: in general we tend not to want code that doesn't do anything. We can always add it later :)

TrueBrain commented on src/lang/english.txt:
```diff
@@ -2026,6 +2026,7 @@ STR_NETWORK_SERVER_LIST_SERVER_VERSION                          :{SILVER}Server
 STR_NETWORK_SERVER_LIST_SERVER_ADDRESS                          :{SILVER}Server address: {WHITE}{RAW_STRING}
 STR_NETWORK_SERVER_LIST_START_DATE                              :{SILVER}Start date: {WHITE}{DATE_SHORT}
 STR_NETWORK_SERVER_LIST_CURRENT_DATE                            :{SILVER}Current date: {WHITE}{DATE_SHORT}
+STR_NETWORK_SERVER_LIST_GAMESCRIPT                              :{SILVER}Gamescript: {WHITE}{RAW_STRING} {SILVER}Version: {WHITE}{COMMA}
```
I am a bit on the fence if the version is useful, but either way, I would do: `{RAWSTRING} (v{NUM})` instead.

TheDude-gh replied:
I wanted to use {NUM} actually, but then I saw that client count is {COMMA}, so I choose to keep it same, although {NUM} is fine.

TrueBrain commented on src/network/core/game_info.cpp:
```diff
@@ -198,6 +200,21 @@ void SerializeNetworkGameInfo(Packet *p, const NetworkServerGameInfo *info)
+	/* NETWORK_GAME_INFO_VERSION = 5 */
+	{
+		/* Only send GameScript details if server is running some Gamescript */
+		GameInfo* gsinfo = Game::GetInfo();
```
Coding-style: `*` should be on the right side, not the left.

TheDude-gh replied:
Actually I don't know what do you mean by that.

TrueBrain replied:
`GameInfo *gsinfo` is our coding style. `*` on the other side of the space :)

TrueBrain commented on src/network/core/game_info.cpp:
```diff
@@ -22,6 +22,8 @@
 #include "../network.h"
 #include "packet.h"
+#include "../../game/game.hpp"
+#include "../../game/game_info.hpp"
```
I think only either of these two are required, not?

TheDude-gh replied:
Well, game_info.hpp could be omitted and the code changed from

GameInfo* gsinfo = Game::GetInfo();
p->Send_uint32((uint32)gsinfo->GetVersion());

to

p->Send_uint32((uint32)Game::GetInfo()->GetVersion());

So, whichever you prefere.

TrueBrain replied:
No, those blobs do exactly the same. The fact that you do not assign it to a variable doesn't change anything ;) In both cases it needs to know `GameInfo` class. So those two variants really are identical, as far as the compiler cares.

But I checked, both headers are needed. I assumed `game.hpp` would include `game_info.hpp`, but it uses `class GameInfo` instead. That is fair, so what you have is fine.

Headers in OpenTTD are wildly inconsistent in how they work, so sometimes it is a bit surprising :)

TrueBrain reviewed (commented):
A lot better, tnx :) Now it also makes a lot more sense what you are doing :D

TheDude-gh commented:
One thing that bothers me with adding gamesript info to game info packet, and the reason I haven't do that in the older fork, is that it breaks getting game info for any older openttd versions before this patch (if it would be included).

For some reason I do not quite understand new stuff in info packet is added before the old. So when older version reads info packet from new hypothetical version, it will read GS info where it would currently expect NewGRF info, and thus fail.

Solution would be to start adding any new info data to the end of the packet, so when older version reads new version packet, it reads what it knows and throw away the rest. But then it would not be consistent with current coding choice.

What do we about that? I myself would prefer functionality of older versions before consistency of the code.

TrueBrain commented:
>
> What do we about that? I myself would prefer functionality of older versions before consistency of the code.

Partial true what you mention. What happens is this:

Older clients that request data from a newer servers, get "v5". They do not know how to read it, so skip reading it completely (they never hit any case in the switch).
Newer clients that request data from an older server just starts at "v4", and can happily read it.

This is mostly always the case with protocol changes. It is backwards compatible, but not forwards compatible.

That said, we are currently reworking some parts of the networking code, among which the GameInfo. In #9017 it is already bumping GameInfo to v5, and this can just be part of that bump. So v5 is already happening, so there is no reason to resolve it in another way :)

TheDude-gh commented:
True, it will not run into any problems, but it will still display the new version unreadable info in GUI like this:

![lobby](https://user-images.githubusercontent.com/47122855/117861973-5ec6fa80-b292-11eb-9d32-8995b6cdc474.png)

So old versions will display lot of this empty rows when there are many new version servers. 
I would stick to "people should upgrade their software" motto.

TheDude-gh committed:
String change: version from {COMMA} to {NUM}
Removing useles code
Packet sends only version and GS name
Change width of details right pane in lobby window (WID_NG_DETAILS) from 140 to 270px, so even longer strings are visible

TrueBrain commented:
Tnx!

Most likely I cherry-pick this PR into #9017, or rebase it after that is merged. As that is already bumping game-info version, might as well make it part of the package. But that needs a bit more time, so stay tuned :)

TrueBrain commented:
As promised, I integrated this in #9017. Will close this PR once that PR is merged :)
